core/src/main/scala/kafka/controller/ControllerEventManager.scala

```diff
@@ -41,7 +43,14 @@ class ControllerEventManager(rateAndTimeMetrics: Map[ControllerState, KafkaTimer
 
   def close(): Unit = thread.shutdown()
 
-  def put(event: ControllerEvent): Unit = queue.put(event)
+  def put(event: ControllerEvent): Unit = inLock(putLock) {
+    queue.put(event)
+  }
+
+  def clearAndPut(event: ControllerEvent): Unit = inLock(putLock) {
+    queue.clear()
```

Comment:
doWork() doesn't use putLock.
Is it possible that an event retrieved by doWork() is supposed to be cleared by this call ?

Reply:
I think it's a good point that we should document the expected behaviour even if there's no bug. @onurkaraman, can we please do that in a follow-up?

Reply:
The purpose of putLock is to make sure no other callers can put anything to the queue in the middle of a clearAndPut() call. It's ok for a reader to have taken an event out of the queue just before the queue is cleared since in KafkaController.expire(), we wait until the last event in the queue is processed before creating a new ZK session. We can probably document this to make it clear.